### PR TITLE
Pi Zero Compatibility; DTB parser

### PR DESCRIPTION
**Pull request by fayalalebrun**
- I added the option to run on the raspberry pi zero QEMU board quite easily, by adding a BOARD make option. I also removed the CPU and MEMORY options  from the documentation, as they both must be set to a specific value for each board anyways, making them redundant.
- As I have been discussing in #76 I've been looking for ways to detect the hardware information in the best way possible. Since QEMU doesn't provide the dtb file in r2 when starting from a .elf file, I decided to link it in the linker file, depending on the board that is passed to make. This doesn't mean that the kernel is compiled for a specific board, as we can still use the dtb that the bootloader passes to us in real hardware. This is just to circumvent that limitation in QEMU.

### CODE REVIEW

**Comment by fayalalebrun** [comment]
The OS now runs on the Raspberry Pi Zero and is passing all tests. This is what has been added.

- A parser which performs a search for a specified property in a DTB. It does this without allocating any data structures, making it well suited for running when OS initialization is in its early stages. This is now used for board and memory detection.
- The UART driver was refactored, so it may be used for both the BCM2835 and BCM2836, as it is identical in both.
- New timer driver for use with the BCM2835, which uses the BCM's system timer. The timer which was previously used, which was interfaced with via the coprocessor, is not available in the ARM1176, which made this necessary.

**Comment by fayalalebrun** [comment]
Resolves #56, resolves #76

File: src/kernel/src/common/start.c [hunk 1]
```diff
@@ -39,32 +44,51 @@ void init() {
     asm volatile("pop {lr}");
 }
 
+
 /// Entrypoint for the C part of the kernel.
 /// This function is called by the assembly located in [startup.s].
 /// The MMU has already been initialized here but only the first MiB of the kernel has been mapped.
-void start(uint32_t * p_bootargs, size_t memory_size) {
+void start(uint32_t * p_bootargs, struct DTHeader * dtb) {
```
Comment (jonay2000) [inline src/kernel/src/common/start.c:51]: maybe not do this in start.c (including the global variable). Might be better to move that to a seperate file which which exports a function `get_DTB()` or similar. `start.c` is nicer if it's jut a large list of function calls to set up the system, and preferably has as little control flow as possible.

File: src/kernel/src/common/start.c [hunk 2]
```diff
@@ -39,32 +44,51 @@ void init() {
...
     // After this point, code can request the hardware info struct to find out what
     // Code should be ran.
-    init_hardwareinfo();
+    init_hardwareinfo(dtb);
 
     // Initialize the chipset and enable uart
     init_chipset();
 
-    INFO("Detected memory size: 0x%x Bytes", memory_size);
+    struct DTProp * mem_prop = dtb_get_property(dtb, "/memory", "reg");
```
Comment (jonay2000) [inline src/kernel/src/common/start.c:65]: same here

File: src/kernel/src/common/startup.s [hunk 3]
```diff
@@ -6,14 +6,11 @@ _Reset:
     // Disable other cores
     mrc p15, #0, r1, c0, c0, #5
     and r1, r1, #3
-    cmp r1, #0
+    cmp r1, #3
```
Comment (jonay2000) [inline src/kernel/src/common/startup.s:9]: what's the advantage of 3 here?

**Review by jonay2000 — CHANGES_REQUESTED** [review]
Are all the binary blobs really necessary? I believe you're only using 2 of them. Maybe make a small index in `src/kernel/dtb/README.md` explaining the use of the files that are there and maybe even some more documentation about how the DTBs work (even if it's just links to the relevant docs). That could help future groups a lot.

File: src/kernel/src/common/start.c (hunk 1 above)
Comment (fayalalebrun) [reply src/kernel/src/common/start.c:48]: I've moved most of this to hardware info. However, do you mean that I should not pass the dtb header pointer to start? How else could I do it? That pointer must go from the bootloader to startup.s to start.c

File: src/kernel/src/common/startup.s (hunk 3 above)
Comment (fayalalebrun) [reply src/kernel/src/common/startup.s:9]: Though I have not been able to obtain a multi-core Raspberry Pi to verify, everything indicates that newer versions of the RPi bootloader disable all cores except one upon entry. However, in older bootloader versions and in QEMU this is not the case. Older versions of the linux kernel which had to deal with multiple cores being active act as you see here.

The co-processor register that is being read is not documented on the ARM1176. However, it happens to return this value.

Without this change, as you can imagine an Rpi Zero would be stuck forever.

**Pull request merged**